### ISSUE TITLE
Add ability to prevent global variables definition

### ISSUE THREAD

**Issue by Shutnik** [eslint/eslint#4720]
Global variables is bad practice. Sometimes you can forgot to wrap variables into IIFE if you don't actually want them global.

For example

```js
if (DEBUG) {
  var consoleError = console.error; // global variable
  console.error = function () {
    // do something
    console.error.apply(console, arguments);
  }
}
```

instead of

```js
if (DEBUG) {
  (function() {
    var consoleError = console.error; // no global wariable
    console.error = function () {
      // do something
      console.error.apply(console, arguments);
    }
  )()};
}
```

Proposed core rule testcase:

```js
/*exported a, b*/

var a = 0; // no error since exported

let b = 1; // no error since exported

var c = 2; // error since global

let d = 3; // error since local for global scope

if (DEBUG) {
  var e = 4; // error since global

  let f = 5; // no error since local for scope
}

function () {
  var g = 6; // no error since local for function

  let h = 7; // no error since local for function

  if (DEBUG) {
    var i = 8; // no error since local for function

    let k = 9; // no error since local for scope
  }
}
```

**Comment by eslintbot** [comment]
@Shutnik Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by nzakas** [comment]
We are adding this in 2.0.0: http://eslint.org/docs/2.0.0/rules/no-implicit-globals
